applications/GeoMechanicsApplication/tests/cpp_tests/test_time_incrementor.cpp

```diff
@@ -0,0 +1,439 @@
+}
+
+KRATOS_TEST_CASE_IN_SUITE(AdaptiveTimeIncrementorThrowsWhenMaxNumberOfIterationsEqualsZero, KratosGeoMechanicsFastSuite)
+{
+    const auto start_time            = 0.0;
+    const auto end_time              = start_time + 8.0;
+    const auto start_increment       = 0.5;
+    const auto max_num_of_cycles     = std::size_t{8};
+    const auto reduction_factor      = 0.5;
+    const auto increase_factor       = 2.0;
+    const auto min_num_of_iterations = std::size_t{0};
+    const auto max_num_of_iterations = std::size_t{0};
+
+    KRATOS_EXPECT_EXCEPTION_IS_THROWN(AdaptiveTimeIncrementor(start_time, end_time, start_increment, max_num_of_cycles, reduction_factor, increase_factor, min_num_of_iterations, max_num_of_iterations),
+                                      "Minimum number of iterations (0) is not less than maximum number of iterations (0)")
+}
+
+KRATOS_TEST_CASE_IN_SUITE(AdaptiveTimeIncrementorThrowsWhenMinNumberOfIterationsIsGreaterThanMaxNumberOfIterations, KratosGeoMechanicsFastSuite)
+{
+    const auto start_time            = 0.0;
+    const auto end_time              = start_time + 8.0;
+    const auto start_increment       = 0.5;
+    const auto max_num_of_cycles     = std::size_t{8};
+    const auto reduction_factor      = 0.5;
+    const auto increase_factor       = 2.0;
+    const auto max_num_of_iterations = std::size_t{10};
+    const auto min_num_of_iterations = max_num_of_iterations + 1;
+
+    KRATOS_EXPECT_EXCEPTION_IS_THROWN(AdaptiveTimeIncrementor(start_time, end_time, start_increment, max_num_of_cycles, reduction_factor, increase_factor, min_num_of_iterations, max_num_of_iterations),
+                                      "Minimum number of iterations (11) is not less than maximum number of iterations (10)")
+}
+
+KRATOS_TEST_CASE_IN_SUITE(WantNextTimeStepWhenEndTimeIsNotAttainedYet, KratosGeoMechanicsFastSuite)
+{
+    const auto start_time        = 0.0;
+    const auto end_time          = start_time + 8.0;
+    const auto start_increment   = 0.5;
+    const auto time_incrementor  = AdaptiveTimeIncrementor{start_time, end_time, start_increment};
+    auto       previous_state    = TimeStepEndState{};
+    previous_state.time          = end_time - 1.0;
+
+    KRATOS_EXPECT_TRUE(time_incrementor.WantNextStep(previous_state))
+}
+
+KRATOS_TEST_CASE_IN_SUITE(NoNextTimeStepWhenEndTimeIsExceeded, KratosGeoMechanicsFastSuite)
+{
+    const auto start_time        = 0.0;
+    const auto end_time          = start_time + 8.0;
+    const auto start_increment   = 0.5;
+    const auto time_incrementor  = AdaptiveTimeIncrementor{start_time, end_time, start_increment};
+    auto       previous_state    = TimeStepEndState{};
+    previous_state.time          = end_time + 1.0;
+
+    KRATOS_EXPECT_FALSE(time_incrementor.WantNextStep(previous_state))
+}
+
+KRATOS_TEST_CASE_IN_SUITE(NoNextTimeStepWhenAtEndTime, KratosGeoMechanicsFastSuite)
+{
```

Comment:
I think we just miss one test case, where the previous_state is before the end time and previous_state + start_increment is after the end time (i.e. the last step would push it over the end time)

Reply:
I may misunderstand you, but I believe this is test case `ReduceIncrementToAvoidExceedingEndTime`. Could you please verify whether that one covers what you had in mind? Otherwise, let's discuss what you meant and add a corresponding unit test. Thanks!

Reply:
So in that case, the 'WantNextStep' function should return true? Maybe we should add that assert to the `ReduceIncrementToAvoidExceedingEndTime` test then? Just checking if this specific function returns the correct value in that case